accounts/tests.py

```diff
@@ -156,3 +159,46 @@ def test_user_refresh_token(self):
         refresh_response = self.client.post(self.token_refresh_url, self.user_credential, format='json')
 
         self.assertNotEqual(token, refresh_response.data['token'])
+
+
+class SignupTest(APITestCase):
+
+    def setUp(self):
```

Comment:
If this is your full unittest, please cover a few invalid username/password cases, e.g. empty username, username too long(https://www.google.com/search?q=username+length&oq=username+length&aqs=chrome..69i57j0l5.4219j0j7&sourceid=chrome&ie=UTF-8), password should meet policies (https://en.wikipedia.org/wiki/Password_policy). Our DB will have constraints but we don't want to always insert invalid username/password and then let DB throw error. Our backend service should immediately return error which is faster.

Reply:
The username and password validations are tested in the test case class - AccountsTest, which I believe already satisfied the cases you mentioned. I agree with you most validation shouldn't hit the DB. Basically only the uniqueness validation needs to hit the DB; others are handled by the request data serializer (see `accounts/serializers.py`.) The complete list of tests are:

1. test_create_user_return_valid_token
1. test_create_user_with_short_password
1. test_create_user_with_no_password
1. test_create_user_with_too_long_username
1. test_create_user_with_no_username
1. test_create_user_with_preexist_username - hits DB
1. test_create_user_with_preexist_email - hits DB
1. test_create_user_with_invalid_email
1. test_create_user_with_no_email